pkg/objstore/oss/oss.go

```diff
@@ -0,0 +1,366 @@
+
+type Bucket struct {
+	name   string
+	logger log.Logger
+	client *alioss.Client
+	config Config
+	bucket *alioss.Bucket
+}
+
+func NewTestBucket(t testing.TB) (objstore.Bucket, func(), error) {
+	c := configFromEnv()
+	err := func(conf Config) error {
+		if conf.Endpoint == "" {
+			return errors.New("no aliyun oss endpoint in config file")
+		}
+		if conf.AccessKeyID == "" {
+			return errors.New("access_key_id is not present in config file")
+		}
+		if conf.AccessKeySecret == "" {
+			return errors.New("access_key_secret is not present in config file")
+		}
+		return nil
+	}(c)
+	if err != nil {
+		return nil, nil, err
+	}
+	if c.Bucket != "" && os.Getenv("THANOS_ALLOW_EXISTING_BUCKET_USE") == "true" {
+		t.Log("ALIYUNOSS_BUCKET is defined. Normally this tests will create temporary bucket " +
+			"and delete it after test. Unset ALIYUNOSS_BUCKET env variable to use default logic. If you really want to run " +
+			"tests against provided (NOT USED!) bucket, set THANOS_ALLOW_EXISTING_BUCKET_USE=true.")
+		return NewTestBucketFromConfig(t, c, true)
+	}
+	return NewTestBucketFromConfig(t, c, false)
+}
+
+// Upload the contents of the reader as an object into the bucket.
+func (b *Bucket) Upload(ctx context.Context, name string, r io.Reader) error {
+	chunksnum, lastslice := func(string, io.Reader) (int, int64) {
+		switch r.(type) {
+		case *os.File:
+			f, _ := r.(*os.File)
+			if fileInfo, err := f.Stat(); err == nil {
+				s := fileInfo.Size()
+				return int(math.Floor(float64(s) / alioss.MaxPartSize)), s % alioss.MaxPartSize
+			}
+		case *strings.Reader:
+			f, _ := r.(*strings.Reader)
+			return int(math.Floor(float64(f.Size()) / alioss.MaxPartSize)), f.Size() % alioss.MaxPartSize
+		}
+		return -1, 0
+	}(name, r)
+
+	ncloser := ioutil.NopCloser(r)
+	switch chunksnum {
+	case -1:
+		return errors.New("can not guess the size of io.Reader")
+	case 0:
+		if err := b.bucket.PutObject(name, ncloser); err != nil {
+			return errors.Wrap(err, "upload oss object")
+		}
+	default:
+		{
+			init, err := b.bucket.InitiateMultipartUpload(name)
+			if err != nil {
+				return errors.Wrap(err, "initiate multi-part upload failed")
+			}
+			chunk := 0
+			uploadEveryPart := func(everypartsize int64, cnk int) (alioss.UploadPart, error) {
+				return b.bucket.UploadPart(init, ncloser, everypartsize, cnk)
+			}
+			var parts []alioss.UploadPart
+			for ; chunk < chunksnum; chunk++ {
+				part, err := uploadEveryPart(alioss.MaxPartSize, chunk+1)
+				if err != nil {
+					if err := b.bucket.AbortMultipartUpload(init); err != nil {
+						return errors.Wrap(err, "multi-part upload ends with aborting")
+					}
+					return errors.Wrap(err, "multi-part upload one chunk failed")
+				}
+				parts = append(parts, part)
+			}
+			if lastslice != 0 {
+				part, err := uploadEveryPart(lastslice, chunksnum+1)
+				if err != nil {
+					if err := b.bucket.AbortMultipartUpload(init); err != nil {
+						return errors.Wrap(err, "multi-part upload ends with aborting")
```

Comment:
Duplicated **AbortMultipartUpload** here.

Reply:
No,It is not duplicated ,  both norm chunk and the last part should abort muliti-part uploading because of failure.